// @react-theming/storybook-addon/dist/manager/components/ThemeBrowser.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var styled = _interopRequireWildcard(require("./ThemeBrowser.styled"));

var _Toolbar = _interopRequireDefault(require("../UI/Toolbar"));

var _Caption = _interopRequireDefault(require("../UI/Caption"));

var _IconButton = _interopRequireDefault(require("../UI/IconButton"));

var _Text = _interopRequireDefault(require("../UI/Text"));

var _clipboard = require("../../utils/clipboard");

var _useEditors = require("../editors/useEditors");

var _default2 = require("../../utils/default");

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var showThemePath = function showThemePath(selectedValue, fieldSnippetFn) {
  if (!selectedValue) return 'Select value';

  try {
    var fn = fieldSnippetFn || _default2.defaultSnippet;
    return fn(selectedValue);
  } catch (err) {
    return 'try to select value';
  }
};

var ThemeBrowser = function ThemeBrowser(_ref) {
  var theme = _ref.theme,
      isSbDark = _ref.isSbDark,
      selectValue = _ref.selectValue,
      selectWord = _ref.selectWord,
      selectedValue = _ref.selectedValue,
      updateTheme = _ref.updateTheme,
      fieldSnippetFn = _ref.fieldSnippetFn;
  var editors = (0, _useEditors.useEditors)(_useEditors.initialEditors);
  var footerAction = showThemePath(selectedValue, fieldSnippetFn);

  var handlerChange = function handlerChange(value) {
    return updateTheme(value.json);
  };

  return /*#__PURE__*/_react.default.createElement(styled.Container, null, /*#__PURE__*/_react.default.createElement(_Toolbar.default, null, /*#__PURE__*/_react.default.createElement(_Caption.default, null, "Editor:"), /*#__PURE__*/_react.default.createElement(styled.ButtonsEditor, {
    isDark: isSbDark
  }, editors.editorButtons.map(function (btn) {
    return /*#__PURE__*/_react.default.createElement("button", {
      key: btn.name,
      type: "button",
      className: btn.isSelected ? 'active' : '',
      onClick: btn.select
    }, btn.title);
  }))), /*#__PURE__*/_react.default.createElement(styled.ThemeHolder, null, editors.renderCurrentEditor({
    isDark: isSbDark,
    theme: theme,
    onChange: handlerChange,
    selectValue: selectValue,
    selectWord: selectWord
  })), footerAction ? /*#__PURE__*/_react.default.createElement(_Toolbar.default, {
    footer: true
  }, /*#__PURE__*/_react.default.createElement(_IconButton.default, {
    isDark: isSbDark,
    icon: "copy",
    title: "copy to clipboard",
    onClick: (0, _clipboard.copyToClipboard)(footerAction)
  }), /*#__PURE__*/_react.default.createElement(_Text.default, null, footerAction)) : null);
};

var _default = ThemeBrowser;
exports.default = _default;
