// @focus-reactive/react-yaml/dist/extensions/error-stripe.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.errorStripe = errorStripe;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _view = require("@codemirror/view");

var _rangeset = require("@codemirror/rangeset");

/* eslint-disable no-param-reassign */

/* eslint-disable no-restricted-syntax */
var baseTheme = _view.EditorView.baseTheme({
  '.cm-errorStripe': {
    backgroundColor: '#ff390040 !important'
  }
});

var stripe = _view.Decoration.line({
  attributes: {
    "class": 'cm-errorStripe'
  }
});

function stripeDeco(view, position) {
  var line = view.state.doc.lineAt(position);
  var builder = new _rangeset.RangeSetBuilder();

  if (position) {
    builder.add(line.from, line.from, stripe);
  }

  return builder.finish();
}

var showStripes = function showStripes(getErrorPos) {
  return _view.ViewPlugin.fromClass( /*#__PURE__*/function () {
    function _class2(view) {
      (0, _classCallCheck2["default"])(this, _class2);
      (0, _defineProperty2["default"])(this, "decorations", _view.DecorationSet);
      this.decorations = stripeDeco(view, 0);
    }

    (0, _createClass2["default"])(_class2, [{
      key: "update",
      value: function update(_update) {
        if (_update.docChanged || _update.viewportChanged) {
          var _getErrorPos = getErrorPos(_update.state.doc.toString()),
              position = _getErrorPos.position;

          if (!position) {
            this.decorations = stripeDeco(_update.view, 0);
            return;
          }

          this.decorations = stripeDeco(_update.view, position);
        }
      }
    }]);
    return _class2;
  }(), {
    decorations: function decorations(v) {
      return v.decorations;
    }
  });
};

function errorStripe() {
  var getErrorPos = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : function () {
    return {};
  };
  return [baseTheme, showStripes(getErrorPos)];
}
